fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java

```diff
@@ -1241,21 +1242,56 @@ private void checkIfBlockingInFuture(CompletableFuture<?> future) {
         // we will simply log the fact that this has taken place and where it has taken place here.
         StackTraceElement possiblyAsyncReturningLocation = null;
 
-        for (StackTraceElement stackElement : stack) {
+        for (int pos = 0; pos < stack.length; pos++) {
+            StackTraceElement stackElement = stack[pos];
             if (stackElement.getClassName().startsWith(CompletableFuture.class.getName())) {
-                logOrThrowBlockingInAsync(behavior, isComplete, stackElement, BLOCKING_IN_ASYNC_CONTEXT_MESSAGE);
+                //look to see if we can ignore this because of careful annotations
+                boolean canIgnore = false;
+                for (int annotationPos = pos - 1; annotationPos >= 0; annotationPos--) {
+                    StackTraceElement possibleIgnore = stack[annotationPos];
+                    if (canIgnoreBlockingInAsync(possibleIgnore.getClassName(), possibleIgnore.getMethodName())) {
+                        canIgnore = true;
+                        break;
+                    }
+                }
+                if (!canIgnore) {
+                    logOrThrowBlockingInAsync(behavior, isComplete, stackElement, BLOCKING_IN_ASYNC_CONTEXT_MESSAGE);
+                }
             } else if (stackElement.getMethodName().endsWith("Async")) {
                 possiblyAsyncReturningLocation = stackElement;
             }
         }
-
         if (possiblyAsyncReturningLocation != null && !isComplete) {
             // Maybe one day this will be configurable, but for now we will only allow this situation to log
             logOrThrowBlockingInAsync(BlockingInAsyncDetection.IGNORE_COMPLETE_WARN_BLOCKING, isComplete,
                     possiblyAsyncReturningLocation, BLOCKING_RETURNING_ASYNC_MESSAGE);
         }
     }
 
+    private boolean canIgnoreBlockingInAsync(final String className, final String methodName) {
```

Comment:
This should probably have a test added in `FDBDatabaseTest` (there is a suite of tests around this named `testBlockingInAsync*`